File: Sort/Quick_Sort.cpp
#include <stdio.h>
 
void swap(int* a, int i, int j){
	int temp = a[i];
	a[i] = a[j];
	a[j] = temp;
}
// 切分操作 
int partition(int* a, int low, int high){
	int i = low;
	// 下面用的--j，必须这么写，改成j--不行 
	int j = high + 1;
	// 选最左为切分元素 
	int v = a[low];
	
	while(true){
		while(a[++i] < v) if(i == high) break;
		while(a[--j] > v) if(j == low) break; 
		if(i >= j) break;
		swap(a, i, j);
	}
	swap(a, low, j);
	
	return j;
} 
// high = n-1;
void sort(int* a,int low, int high){
	if(low >= high) return;
	int j = partition(a, low, high);
	sort(a, low, j-1);
	sort(a, j+1, high);
} 

int main(int argc, char** argv){
	int a[10] = {0, 4, 2, 5, 12, 23, 4, 3, 6, 1};
	sort(a, 0, 9);
	for(int i = 0; i<10; i++)
		printf("%d\n", a[i]);
}

File: Sort/quickSort.py
# -*- coding: UTF-8 -*-
# 快速排序。
import numpy as np


def swap(lyst, i, j):
    lyst[i], lyst[j] = lyst[j], lyst[i]


def partition(lyst, low, high):
    i = low
    j = high
    pivot = lyst[low]

    while True:
        # 2个while都要 = 否则出现 [a, a, ..., a]会无限循环
        while lyst[i] <= pivot:
            i += 1
            if i == high: break

        while lyst[j] >= pivot:
            j -= 1
            if j == low: break

        if i >= j: break

        swap(lyst, i, j)

    swap(lyst, low, j)

    return j


def quickSortHelper(lyst, low, high):
    if low >= high:
        return
    pivotLocation = partition(lyst, low, high)
    quickSortHelper(lyst, low, pivotLocation - 1)
    quickSortHelper(lyst, pivotLocation + 1, high)


def quickSort(lyst):
    quickSortHelper(lyst, 0, len(lyst) - 1)


if __name__ == "__main__":
    lyst = np.random.randint(10, size=10)
    print lyst
    quickSort(lyst)
    print lyst
